src/backend/distributed/commands/multi_copy.c

```diff
@@ -1195,6 +1234,282 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
+	}
+	PG_CATCH();
+	{
+		/* roll back all transactions */
+		EndRemoteCopy(shardConnections->connectionList, false);
+		AbortRemoteTransactions(shardConnections->connectionList);
+		CloseConnections(shardConnections->connectionList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	if (completionTag != NULL)
+	{
+		snprintf(completionTag, COMPLETION_TAG_BUFSIZE,
+				 "COPY " UINT64_FORMAT, processedRowCount);
+	}
+}
+
+
+/*
+ * StartCopyToNewShard creates a new shard and related shard placements, opens
+ * connections to shard placements, and starts copy command by sending copy
+ * headers.
+ */
+static void
+StartCopyToNewShard(CopyOutState copyOutState, ShardConnections *shardConnections,
+					Oid relationId, CopyStmt *copyStatement)
+{
+	char *relationName = get_rel_name(relationId);
+	text *relationNameText = cstring_to_text(relationName);
+	Datum relationNameDatum = PointerGetDatum(relationNameText);
+	Datum shardIdDatum = DirectFunctionCall1(master_create_empty_shard,
+											 relationNameDatum);
+
+	int64 shardId = DatumGetInt64(shardIdDatum);
+	shardConnections->shardId = shardId;
+
+	list_free_deep(shardConnections->connectionList);
+	shardConnections->connectionList = NIL;
+
+	OpenCopyTransactions(copyStatement, shardConnections);
+
+	/* send binary headers to shard placements */
+	resetStringInfo(copyOutState->fe_msgbuf);
+	AppendCopyBinaryHeaders(copyOutState);
+	SendCopyDataToAll(copyOutState->fe_msgbuf,
+					  shardConnections->connectionList);
+}
+
+
+/*
+ * FinalizeCopyToNewShard sends copy footers, commits copy transaction, and
+ * closes connections to shard placements. Finally, it calculates shard
+ * statistics and updates metadata.
+ */
+static void
+FinalizeCopyToNewShard(CopyOutState copyOutState, ShardConnections *shardConnections,
+					   Oid relationId, uint64 processedRowCount)
+{
+	/* send binary footers to all shard placements */
+	resetStringInfo(copyOutState->fe_msgbuf);
+	AppendCopyBinaryFooters(copyOutState);
+	SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+
+	/* close the COPY input on all shard placements */
+	EndRemoteCopy(shardConnections->connectionList, true);
+
+	CommitRemoteTransactions(shardConnections->connectionList, true);
+	CloseConnections(shardConnections->connectionList);
+
+	/* if no rows are copied, drop empty shard */
+	if (processedRowCount != 0)
+	{
+		UpdateShardStatistics(relationId, shardConnections->shardId);
+	}
+	else
+	{
+		char *relationName = get_rel_name(relationId);
+
+		Oid schemaId = get_rel_namespace(relationId);
+		char *schemaName = get_namespace_name(schemaId);
+
+		ShardInterval *shardInterval = LoadShardInterval(shardConnections->shardId);
+		List *shardIntervalList = list_make1(shardInterval);
+
+		DropShards(relationId, schemaName, relationName, shardIntervalList);
+	}
+}
+
+
+/*
+ * UpdateShardStatistics updates metadata for the given shard id.
+ */
+static void
+UpdateShardStatistics(Oid relationId, int64 shardId)
```

Comment:
Are there other places where we do this? Is it possible to use the same logic there? Also wondering whether this is the right place.